### PR TITLE
Refreshtoken

### LINKED ISSUE #23: Handling of jwt, refreshtoken and possibly protected routes
### Including, but not limited to:
- [x] jwt in the app
   - [x] Find a way to store it: https://docs.expo.io/versions/latest/sdk/securestore/
   - [x] Place token on each request in frontend. Can be done via this Link! https://www.apollographql.com/docs/link/links/context/ , if you use this, you can get token out again via context in backend. Described below. 
   - [x] When user logs in, place token in Secure Storage
   - [x] When sending requests, instead of DUMMY TOKEN, grab token from storage. Just comment it in. The code is there, just waiting for the step above to be complete. 
- [x] In backend - Place token on Context, so Resolvers can use it. Don't check here if valid, only check if the token is valid at the protected routes level. Not necessary before. https://www.apollographql.com/docs/apollo-server/security/authentication/#putting-user-info-on-the-context
    - [x] check for valid token - Use AuthenticationError if user is not valid
    - [x] Actually properly check for valid token via jwt.verify by using our secret. Do this when token is properly grabbed and placed in securestorage in frontend upon frontend login
    - [x] protected routes - use ForbiddenError if user is not authorized - Example on getUser resolver.

- [ ] How do we handle refreshtoken? You can check if a token is valid, but just expired, via an options object like {ignoreExpiration:true} as the third argument to jwt.verify. 
    - [ ] Either send an error back, if token is expired, which can then be handled like: Use ErrorLink, so that if unauthenticated, you can send an automatic request for a new token from backend, then redo the request. [via this pattern. ](https://www.apollographql.com/docs/link/links/error/#retrying-failed-requests). Then we'd also need a new endpoint that simply takes an expired token, and returns a new one. 
    - [ ] Or on the backend, use the refreshtoken to get a new accesstoken to the user. 
- [x] Right now we control the logged in state via a useState. And that is fine. But, since user might has a JWT in the SecureStore, we could make a useEffect, that checks for a token in securestore, and if it is there, then decodes it, and logs the user in? So they keep their logged in, even if they close Expo and the App? :) Could be Cool imo https://github.com/Hold-Krykke/Security_and_FullstackJavascript_Exam/pull/28

### LINKED ISSUE #60: We should reset cache on logout
https://www.apollographql.com/docs/react/networking/authentication/#reset-store-on-logout

### PR DESCRIPTION
**Pull request by MalteMagnussen**
### Will demo tomorrow, then we can merge. 

After a great days pair-programming and debugging with @HrBjarup , we finally did it, and it works on both our phones. Couldn't have done it alone. 

https://www.apollographql.com/docs/link/links/error/#retrying-failed-requests

___
**_This is fixed with an Observables workaround, which is recommended by the developers in the github issue linked below:_**
**Issue**
onError isn't async
**Solution**
https://github.com/apollographql/apollo-link/issues/646

[_Can use button to test_](https://github.com/Hold-Krykke/Security_and_FullstackJavascript_Exam/blob/d974c4d2ef2932c60e187775fe05c9d23efb7a03/FindYourFriends/screens/HomeScreen.js)

Put TOKEN_EXPIRATION=3600 into .env in backend

### CODE REVIEW

File: Backend/src/resolverMap.ts [hunk 1]
```diff
@@ -28,7 +28,7 @@ const jwt = require("jsonwebtoken");
 
 const schema: string = process.env.DATABASE_SCHEMA || "";
 
-const tokenExpirationInSeconds = 3600;
+const tokenExpirationInSeconds = 30; // 3600
```
Comment (MalteMagnussen) [inline Backend/src/resolverMap.ts]: Remember to clean-up before merge.

File: Backend/src/server.ts [hunk 2]
```diff
@@ -33,21 +34,23 @@ app.use(passport.initialize());
 
 //The regular logger needs to be before the router
 
-const tokenExpirationInSeconds = 3600;
+const tokenExpirationInSeconds = 25; //3600
```
Comment (MalteMagnussen) [inline Backend/src/server.ts]: Wait, there is two??

File: Backend/package.json [hunk 3]
```diff
@@ -34,12 +34,13 @@
     "mongodb": "^3.5.7",
     "mysql": "^2.18.1",
     "ncp": "^2.0.0",
+    "node-fetch": "^2.6.0",
```
Comment (MalteMagnussen) [inline Backend/package.json:37]: remember to `npm install`

File: FindYourFriends/screens/MapScreen.js [hunk 4]
```diff
@@ -118,6 +126,55 @@ const MapScreen = (props) => {
   );
 };
 
+const UserInfo = () => {
```
Comment (MalteMagnussen) [inline FindYourFriends/screens/MapScreen.js:129]: Remember to remove before merging. Remove post demo.

**Review by Runi-VN — APPROVED** [review]
Good work. :)

**Review by Runi-VN — APPROVED** [review]

**Review by Castau — APPROVED** [review]

**Pull request merged**